packages/snaps-execution-environments/src/webview/ProxyMessageStream.ts

```diff
@@ -0,0 +1,106 @@
+import type { PostMessageEvent } from '@metamask/post-message-stream';
+import { BasePostMessageStream } from '@metamask/post-message-stream';
+import { isValidStreamMessage } from '@metamask/post-message-stream/dist/utils';
+
+type ProxyMessageStreamArgs = {
+  name: string;
+  target: string;
+  targetOrigin?: string;
+  targetWindow?: Window;
+};
+
+/**
+ * A {@link Window.postMessage} stream.
+ */
+
+export class ProxyMessageStream extends BasePostMessageStream {
+  #name;
+
+  #target;
+
+  #targetOrigin;
+
+  #targetWindow;
+
+  /**
+   * Creates a stream for communicating with other streams across the same or
+   * different `window` objects.
+   *
+   * @param args - Options bag.
+   * @param args.name - The name of the stream. Used to differentiate between
+   * multiple streams sharing the same window object.
+   * @param args.target - The name of the stream to exchange messages with.
+   * @param args.targetOrigin - The origin of the target. Defaults to
+   * `location.origin`, '*' is permitted.
+   * @param args.targetWindow - The window object of the target stream. Defaults
+   * to `window`.
+   */
+  constructor({
+    name,
+    target,
+    targetOrigin = window.location.origin,
+    targetWindow = window,
+  }: ProxyMessageStreamArgs) {
+    super();
+
+    if (
+      typeof window === 'undefined' ||
+      typeof window.postMessage !== 'function'
+    ) {
+      throw new Error(
+        'window.postMessage is not a function. This class should only be instantiated in a Window.',
+      );
+    }
+
+    this.#name = name;
+    this.#target = target;
+    this.#targetOrigin = targetOrigin;
+    this.#targetWindow = targetWindow;
+
+    this._onMessage = this._onMessage.bind(this);
+
+    window.addEventListener(
```

Comment:
This also uses the global, which may be intentional. Should we have both `targetWindow` and `originWindow`?